db/functions/callbacks.go

```diff
@@ -0,0 +1,244 @@
+/*
+Copyright 2022-Present Couchbase, Inc.
+
+Use of this software is governed by the Business Source License included in
+the file licenses/BSL-Couchbase.txt.  As of the Change Date specified in that
+file, in accordance with the Business Source License, use of this software will
+be governed by the Apache License, Version 2.0, included in the file
+licenses/APL2.txt.
+*/
+
+package functions
+
+import (
+	"bytes"
+	"context"
+	"errors"
+	"fmt"
+	"net/http"
+
+	"github.com/couchbase/gocb/v2"
+	sgbucket "github.com/couchbase/sg-bucket"
+	"github.com/couchbase/sync_gateway/base"
+	"github.com/couchbase/sync_gateway/db"
+)
+
+type n1qlUserArgument struct {
+	Name     *string  `json:"name,omitempty"`
+	Channels []string `json:"channels,omitempty"`
+	Roles    []string `json:"roles,omitempty"`
+}
+
+//////// DATABASE EVALUATOR DELEGATE:
+
+// The "real" implementation of EvaluatorDelegate.
+type databaseDelegate struct {
+	db   *db.Database     // The database (with user)
+	ctx  context.Context  // Context for timeouts etc.
+	user *userCredentials // User's info
+}
+
+func (d *databaseDelegate) init(dbc *db.Database, ctx context.Context) error {
+	if ctx == nil {
+		return fmt.Errorf("missing Context")
+	}
+	if err := db.CheckTimeout(ctx); err != nil {
+		return err
+	}
+	d.db = dbc
+	d.ctx = ctx
+	if dbUser := dbc.User(); dbUser != nil {
+		d.user = &userCredentials{
+			Name:     dbUser.Name(),
+			Roles:    dbUser.RoleNames().AllKeys(),
+			Channels: dbUser.Channels().AllKeys(),
```

Comment:
There's something weird here we can potentially address in the future users and roles are collection scoped, but a principal object can see all of them.

The graphql is not collection aware so this isn't immediately problematic.